import: bring TestImportWorkerFailure back online

cucaroach opened the pull request:
## jobs: fix waitForJobToHaveStatus flakey error

Sometimes we flake with the error:

```
 jobs_verification.go:88: error scanning '&{<nil> 0xc032ce2300}': sql: no rows in result set
```

We need this error to be returned and not fail the test in order for the
SucceedsSoon to work. Now if the job fails to be created after 2m the
error will be:

```
import_stmt_test.go:5482: condition failed to evaluate within 2m0s: from jobs_verification.go:90: job 891470100285554693 not found
```

Informs: #107456
Epic: none
Release note: none

## import: allow import worker failure to fail

This commit flushes out work done in #108626 to allow import to succeed
or fail and still have it count as passing.  This covers the case
where the import statement succeeds but the import fails later.

We also handle the case where the import statement reports an error
but actually just puts job in the background.  This is the
"node liveness error: restarting in background" error.

Informs: #107456
Release note: none
Epic: none

## import: unskip TestImportWorkerFailure for stress/race/deadlock

Now that we allow this to fail as long as it fails atomically and it
seems to pass reliably under stress race and deadlock.

Epic: none
Release note: none

cucaroach committed:
jobs: fix waitForJobToHaveStatus flakey error

Sometimes we flake with the error:

```
 jobs_verification.go:88: error scanning '&{<nil> 0xc032ce2300}': sql: no rows in result set
```

We need this error to be returned and not fail the test in order for the
SucceedsSoon to work. Now if the job fails to be created after 2m the
error will be:

```
import_stmt_test.go:5482: condition failed to evaluate within 2m0s: from jobs_verification.go:90: job 891470100285554693 not found
```

Epic: none
Release note: none
Informs: #107456

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/108742)
<!-- Reviewable:end -->

michae2 reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 1 of 1 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/108742#-:-NbqDrbm-ze5Yzd60-mM:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @lidorcarmel and @stevendanna)



<!-- Sent from Reviewable.io -->

stevendanna commented on pkg/testutils/jobutils/jobs_verification.go:
```diff
@@ -82,10 +82,19 @@ func waitForJobToHaveStatus(
-		query := fmt.Sprintf("SELECT status, payload FROM (%s)", InternalSystemJobsBaseQuery)
-		db.QueryRow(t, query, jobID).Scan(&status, &payloadBytes)
+		if !rows.Next() {
+			return errors.Errorf("no rows returned")
```
I suppose we could check rows.Err() first. Not really sure if you'd ever see a non-nil error case on the first row.

stevendanna reviewed (approved):
Thanks!

cucaroach committed:
import: allow import worker failure to fail

This commit flushes out work done in #108626 to allow import to succeed
or fail and still have it count as passing.  This covers the case
where the import statement succeeds but the import fails later.

We also handle the case where the import statement reports an error
but actually just puts job in the background.  This is the
"node liveness error: restarting in background" error.

Informs: #107456
Release note: none
Epic: none

cucaroach commented:
With this in place we now mostly fail with:

```
    import_stmt_test.go:5486: condition failed to evaluate within 2m0s: from jobs_verification.go:79: job 891719423356764164 not found
```

But we have a new exciting failure mode, now we can successfully import everything and all the data is there but the job status will a StatusFailed!

cucaroach committed:
import: unskip TestImportWorkerFailure for stress/race/deadlock

Now that we allow this to fail as long as it fails atomically and it
seems to pass reliably under stress race and deadlock.

Epic: none
Release note: none

celiala commented:
FYI that 23.1.9 is now frozen (I'll remove the backport-23.1.9 label).

the 23.1.10 staging branch is open -- I'll go ahead and add the backport-23.1.10 label in case that is applicable.

stevendanna commented:
@cucaroach Apologies we didn't follow up here.  Is this PR still relevant. I know y'all dug in here and got to the bottom of some things but I'm not clear on the current status of the relevant branches.

cucaroach commented:
> @cucaroach Apologies we didn't follow up here. Is this PR still relevant. I know y'all dug in here and got to the bottom of some things but I'm not clear on the current status of the relevant branches.

I still have some i's to dot on this but I believe it will be test only changes so fine to push back to next train.
